main/core/src/eval/Evaluator.scala

```diff
@@ -598,221 +547,89 @@ case class Evaluator(
-      } else {
-        evalLog.debug(s"Scheduling ${newInProgress.size} new tasks (issuer: ${issuer}): ${newInProgress.map(t => printTerm(t._1)).mkString(", ")}")
-
-        // schedule for parallel execution
-        newInProgress.zipWithIndex.foreach {
-          case (curWork@(terminal, group), index) =>
-            val term = printTerm(terminal)
-            //
-            // Multithreading: Creating the Future
-            //
-            val workerFut: java.util.concurrent.Future[FutureResult] = completionService.submit { () =>
-              if (failFast && state.someTaskFailed.get()) {
-                // we do not start this tasks but instead return with aborted result
-                val newResults = group.map(_ -> Aborted)
-
-                evalLog.debug(s"Skipped evaluation (because of earlier failures): ${term}")
-                FutureResult(curWork, 0, Evaluated(newResults.toMap, Seq(), false))
-
-              } else {
-
-                val counterMsg = state.nextCounterMsg()
-
-                timeLog.debug(s"START [${counterMsg}] ${term}")
-                val startTime = System.currentTimeMillis()
-
-                val res = evaluateGroupCached(
-                  terminal,
-                  group,
-                  state.results.asScala,
-                  counterMsg,
-                  reporter,
-                  testReporter,
-                  logger
-                )
+        def label(x: Terminal) = x match {
+          case Left(x) => x.toString
+          case Right(Labelled(x, y)) => x.toString
+        }
 
-                val endTime = System.currentTimeMillis()
+        val groupGraph = interGroupDeps.items.map { case (k, vs) => (label(k._1), vs.map(t => label(t._1))) }.toMap
+        pprint.log(groupGraph, height=99999)
+        val terminals = sortedGroups.keys.toVector
+        //      println("INITIALIZING promises " + terminals.map(label).mkString(", "))
+        val promises = terminals
+          .map(k => (k, scala.concurrent.Promise[Any]))
+          .toMap
+
+        val taskFutures: Map[Terminal, Future[Any]] = promises.map { case (k, p) => (k, p.future) }
+
+        val results = mutable.LinkedHashMap.empty[Task[_], Result[(Any, Int)]]
+        //      println("INITIALIZING futures")
+        val futures = terminals.map { k =>
+          val deps = interGroupDeps((k, sortedGroups.lookupKey(k))).map(_._1)
+          //          println("REGISTERING " + label(k) + " <- " + deps)
+          Future.sequence(deps.map(taskFutures(_)))
+            .map { upstreamValues =>
+              //              println("START " + label(k) + " " + deps.size)
+              //              println("THREAD " + Thread.currentThread().getName())
+              //              println("DEPS " + deps)
+              val startTime = System.currentTimeMillis()
+
+              val res = evaluateGroupCached(k, sortedGroups.lookupKey(k), results, "lol", _ => None, testReporter, logger)
```

Comment:
I'm re-using the same `evaluateGroupCached` function here, to re-use most of the caching logic with the previous parallel scheduler and the sequential implementation. The interface isn't quite what I want, but it'll do for now I think